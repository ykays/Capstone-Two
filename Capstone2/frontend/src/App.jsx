import React, {useState, useEffect} from 'react';
import './App.css'
import NavBar from './NavBar'
import Map from './Map'
import ParksApi from './api'
import useLocalStorageState from './hooks/useLocalStorageState';
import {BrowserRouter, Routes, Route} from "react-router-dom"
import { jwtDecode } from "jwt-decode";
import UserContext from "./UserContext";
import {fetchUserDetailsFromAPI, removeUser} from "./actions/user"
import { useSelector, useDispatch, shallowEqual } from "react-redux";

function App() {
  const dispatch = useDispatch()
  //const userRedux = useSelector(store=>store.user, shallowEqual)

  const [isLoading, setIsLoading] = useState(true);
  const [token, setToken] = useLocalStorageState("token");
  const [user, setUser] = useState();

  async function registerUser(user) {
    try{
      const userToken = await ParksApi.registerUser(user);
      setToken(userToken.token);
      return { success: true };
    }
    catch(err){
      return {success: false, err}
    }
 
  }

  async function logInUser(user) {
    try {
      const userToken = await ParksApi.logInUser(user);
      setToken(userToken.token);
      return { success: true };
    } catch (err) {
      return { success: false, err };
    }
  }

  function logOutUser() {
    dispatch(removeUser());
    setToken(undefined);
    setUser("");
    
  }

  useEffect(() => {
    if (!token || token===undefined) {
      setIsLoading(false);
      return;
    }
    const decoded = jwtDecode(token);
    dispatch(fetchUserDetailsFromAPI(decoded.username))
    setUser((user) => (decoded.username));
  }, [token]);


  return (
    <>
    <UserContext.Provider value={{ user, setUser }}> 
      <NavBar registerUser={registerUser} logInUser={logInUser} logOutUser={logOutUser}/>
      <Map /> 
    </UserContext.Provider>
    
     
     
    </>
  )
}

export default App
