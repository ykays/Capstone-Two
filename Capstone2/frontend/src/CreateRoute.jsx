import React, { useState, useEffect } from "react";
import ButtonGroup from "@mui/material/ButtonGroup";
import Button from "@mui/material/Button";
import { styled, useTheme } from "@mui/material/styles";
import Box from "@mui/material/Box";
import { TextField } from "@mui/material";
import Drawer from "@mui/material/Drawer";
import CssBaseline from "@mui/material/CssBaseline";
import Divider from "@mui/material/Divider";
import IconButton from "@mui/material/IconButton";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import AddLocationIcon from "@mui/icons-material/AddLocation";
import ParkApi from "./api";
import Tooltip from "@mui/material/Tooltip";
import Alert from "@mui/material/Alert";
import Typography from "@mui/material/Typography";
import { routeReorder } from "./helpers/routeHelper";
import CreateRouteList from "./CreateRouteList";
import Stack from "@mui/material/Stack";

const drawerWidth = 240;

const DrawerHeader = styled("div")(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  padding: theme.spacing(0, 1),
  ...theme.mixins.toolbar,
  justifyContent: "flex-start",
}));

function CreateRoute({
  setNewRoute,
  setAddOnMap,
  setNewRoutePoints,
  newRoutePoints,
  user,
}) {
  const theme = useTheme();
  const [open, setOpen] = useState(false);
  const [routeNameNotes, setRouteNameNotes] = useState({
    routeName: "",
    routeNotes: "",
  });

  const handleDrawerOpen = () => {
    setOpen(true);
    setNewRoute(true);
    setNewRoutePoints([]);
  };

  const handleDrawerClose = () => {
    setOpen(false);
    setNewRoute(false);
  };

  const [msg, setMsg] = useState([]);

  const handleChangeRouteNameNotes = (e) => {
    const { name, value } = e.target;
    setRouteNameNotes((form) => ({ ...form, [name]: value }));
  };

  const handleSaveRoute = async (e) => {
    e.preventDefault();

    const routeName = routeNameNotes["routeName"];
    const routeNotes = routeNameNotes["routeNotes"];
    // const routeName = e.target[0].value;
    // const routeNotes = e.target[2].value;

    const routeDetails = newRoutePoints.map((point) => {
      return {
        waypointName: point[2],
        waypointLongitude: point[1],
        waypointLatitude: point[0],
        parkFlag: point[3] === "park" ? true : false,
        parkCode: point[3] === "park" ? point[4] : null,
      };
    });

    try {
      const result = await ParkApi.createNewRoute(
        user.username,
        routeName,
        routeNotes,
        routeDetails
      );
      routeNameNotes["routeName"] = "";
      routeNameNotes["routeNotes"] = "";
      setNewRoutePoints([]);
      return setMsg({ msg: "Route saved", status: "success" });
    } catch (e) {
      return setMsg({ msg: `Route not saved ${e}`, status: "error" });
    }
  };

  const handleCancelRoute = (e) => {
    routeNameNotes["routeName"] = "";
    routeNameNotes["routeNotes"] = "";
    setNewRoutePoints([]);
    setMsg([]);
  };

  const onDragEnd = ({ destination, source }) => {
    if (!destination) return;

    const newItems = routeReorder(
      newRoutePoints,
      source.index,
      destination.index
    );

    setNewRoutePoints(newItems);
  };

  if (open) {
    return (
      <>
        <Box sx={{ display: "flex" }}>
          <CssBaseline />
          <Drawer
            sx={{
              width: drawerWidth,
              flexShrink: 0,
              "& .MuiDrawer-paper": {
                width: drawerWidth,
              },
            }}
            variant="persistent"
            anchor="right"
            open={open}
          >
            <DrawerHeader>
              <IconButton onClick={handleDrawerClose}>
                {theme.direction === "rtl" ? (
                  <ChevronLeftIcon />
                ) : (
                  <ChevronRightIcon />
                )}{" "}
                <Typography>New Route</Typography>
              </IconButton>
            </DrawerHeader>
            <Divider />

            <Box
              component="form"
              sx={{
                "& .MuiTextField-root": { m: 0.5, width: "25ch" },
              }}
              autoComplete="off"
              display="flex"
              flexDirection="column"
              onSubmit={(e) => {
                handleSaveRoute(e);
              }}
            >
              {msg.length !== 0 && (
                <Alert
                  severity={msg.status}
                  onClose={() => {
                    setMsg([]);
                  }}
                >
                  {msg.msg}
                </Alert>
              )}

              <TextField
                className="required"
                required
                id="outlined-required"
                label="Name"
                size="small"
                name="routeName"
                value={routeNameNotes.routeName}
                inputProps={{ required: true }}
                onChange={handleChangeRouteNameNotes}
              />

              <TextField
                id="outlined"
                label="Route Notes"
                size="small"
                name="routeNotes"
                value={routeNameNotes.routeNotes}
                onChange={handleChangeRouteNameNotes}
              />

              <Divider />
              <Typography style={{ fontSize: "0.6rem" }}>
                To add a Waypoint, click on 'Add Waypoint' and then either click
                on a map or select a park & add to route
              </Typography>
              <Button
                size="small"
                width="10vw"
                color="primary"
                variant="outlined"
                onClick={() => setAddOnMap(true)}
              >
                <AddLocationIcon></AddLocationIcon>Add Waypoint
              </Button>
              <CreateRouteList
                onDragEnd={onDragEnd}
                newRoutePoints={newRoutePoints}
              />
              {newRoutePoints.length > 1 && (
                <ButtonGroup size="small" aria-label="Small button group">
                  {user.length !== 0 ? (
                    <Button type="submit" variant="contained">
                      Save Route
                    </Button>
                  ) : (
                    <Tooltip
                      title="Please log in to save the route"
                      placement="top-start"
                    >
                      <span>
                        <Button variant="contained" disabled>
                          Save Route
                        </Button>
                      </span>
                    </Tooltip>
                  )}
                  <Button
                    variant="outlined"
                    onClick={(e) => handleCancelRoute(e)}
                  >
                    Cancel
                  </Button>
                </ButtonGroup>
              )}
            </Box>
          </Drawer>
        </Box>
      </>
    );
  }

  return (
    <Stack direction="row" sx={{ width: "12%" }}>
      <div>
        <Button onClick={handleDrawerOpen}>Create Route</Button>
      </div>
    </Stack>
  );
}

export default CreateRoute;
