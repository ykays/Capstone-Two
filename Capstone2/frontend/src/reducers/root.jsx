import parks from "./parks";
import filters from "./filters";
import user from "./user";
import routes from "./routes";
import waypoints from "./waypoints";
import { combineReducers } from "redux";

export default combineReducers({
  parks,
  filters,
  user,
  routes,
  waypoints,
});
