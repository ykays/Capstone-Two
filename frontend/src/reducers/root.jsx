import parks from "./parks";
import filters from "./filters";
import user from "./user";
import routes from "./routes";
import { combineReducers } from "redux";

export default combineReducers({
  parks,
  filters,
  user,
  routes,
});
