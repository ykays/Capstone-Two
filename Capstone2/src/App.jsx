import './App.css'

function App() {
  

  return (
    <>
     <h1>Pin the Park App</h1>
    </>
  )
}

export default App
