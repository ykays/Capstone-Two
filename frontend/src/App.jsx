import React, { useState, useEffect } from "react";
import "./App.css";
import NavBar from "./NavBar";
import Map from "./Map";
import ParksApi from "./api";
import useLocalStorageState from "./hooks/useLocalStorageState";
import { jwtDecode } from "jwt-decode";
import { fetchUserDetailsFromAPI, removeUser } from "./actions/user";
import { useDispatch } from "react-redux";

function App() {
  const dispatch = useDispatch();

  const [isLoading, setIsLoading] = useState(true);
  const [token, setToken] = useLocalStorageState("token");

  //function to register user (will be passed to Register Component)
  async function registerUser(user) {
    try {
      const userToken = await ParksApi.registerUser(user);
      setToken(userToken.token);
      return { success: true };
    } catch (err) {
      return { success: false, err };
    }
  }

  //function to register user (will be passed to Login Component)
  async function logInUser(user) {
    try {
      const userToken = await ParksApi.logInUser(user);
      setToken(userToken.token);
      return { success: true };
    } catch (err) {
      console.log(err, "LUAAAERRORAPP");
      return { success: false, err };
    }
  }

  //loging out user (removing from redux store and setting token to undefined in local storage)
  function logOutUser() {
    dispatch(removeUser());
    setToken(undefined);
  }

  // whenever the token changes (register, login, logout) the user details are retrieved and store in redux
  useEffect(() => {
    if (!token || token === undefined) {
      setIsLoading(false);
      return;
    }
    const decoded = jwtDecode(token);
    dispatch(fetchUserDetailsFromAPI(decoded.username));
  }, [token]);

  // setting state for creating new route to be able to manipulate the screen size of entire Map/App
  const [newRoute, setNewRoute] = useState(false);

  return (
    <div style={{ width: newRoute ? "83vw" : "100vw" }}>
      <NavBar
        registerUser={registerUser}
        logInUser={logInUser}
        logOutUser={logOutUser}
      />
      <Map newRoute={newRoute} setNewRoute={setNewRoute} />
    </div>
  );
}

export default App;
