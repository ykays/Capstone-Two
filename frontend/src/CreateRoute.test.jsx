import React, { act } from "react";
import {
  fireEvent,
  render,
  screen,
  waitFor,
  prettyDOM,
} from "@testing-library/react";
import App from "./App";
import CreateRoute from "./CreateRoute.jsx";
import { Provider } from "react-redux";
import root from "./reducers/root.jsx";
import { configureStore } from "@reduxjs/toolkit";

describe("CreateRoute smoke tests", () => {
  const store = configureStore({ reducer: root });
  it("renders the App component without crashing", () => {
    render(
      <Provider store={store}>
        <CreateRoute />
      </Provider>
    );
  });
  it("matches snapshot", function () {
    const { asFragment } = render(
      <Provider store={store}>
        <CreateRoute />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

test("clicking on Create Route", async () => {
  const store = configureStore({ reducer: root });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));

  fireEvent.click(screen.getByText(/Create Route/i));
  expect(screen.getByLabelText(/Name/i)).toBeInTheDocument();
  expect(screen.getByLabelText(/Route Notes/i)).toBeInTheDocument();
  expect(screen.getByText(/To add a Waypoint/i)).toBeInTheDocument();
  expect(screen.getByTestId("AddLocationIcon")).toBeInTheDocument();
});

test("park marker should have Add to route button", async () => {
  const store = configureStore({ reducer: root });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));

  fireEvent.click(screen.getByText(/Create Route/i));
  fireEvent.click(screen.getByRole("button", { name: "acad" }));
  expect(screen.getByText("Add to Route")).toBeInTheDocument();
});

test("park added to route", async () => {
  const store = configureStore({ reducer: root });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));

  fireEvent.click(screen.getByText(/Create Route/i));
  fireEvent.click(screen.getByRole("button", { name: "acad" }));
  fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  expect(screen.getByLabelText("Waypoint")).toHaveAttribute("disabled");
  expect(screen.getByLabelText("Waypoint")).toHaveValue("Acadia National Park");
});

test("creating a route user logged in", async () => {
  const user = {
    username: "testUser",
    email: "testUser@gmail.com",
    firstName: "Test",
    lastName: "User",
  };
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_USER", user: user });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const listItems = screen.getAllByRole("listitem");
  expect(listItems).toHaveLength(2);
  expect(
    screen.getByRole("button", { name: "Save Route" })
  ).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
  expect(
    screen.getByRole("button", { name: "Save Route" })
  ).not.toHaveAttribute("disabled");
});

test("creating a route user not logged in", async () => {
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const listItems = screen.getAllByRole("listitem");
  expect(listItems).toHaveLength(2);
  expect(
    screen.getByRole("button", { name: "Save Route" })
  ).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Save Route" })).toHaveAttribute(
    "disabled"
  );
});

test("park added to the route shoudl have Remove From Route button", async () => {
  const store = configureStore({ reducer: root });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));

  fireEvent.click(screen.getByText(/Create Route/i));
  fireEvent.click(screen.getByRole("button", { name: "acad" }));
  fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  expect(screen.getByLabelText("Waypoint")).toHaveAttribute("disabled");
  expect(screen.getByLabelText("Waypoint")).toHaveValue("Acadia National Park");
  const listItems = screen.getAllByRole("listitem");
  expect(listItems).toHaveLength(1);
  expect(
    screen.getByRole("button", { name: "Remove from Route" })
  ).toBeInTheDocument();
});

test("park removed from the list", async () => {
  const store = configureStore({ reducer: root });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));

  fireEvent.click(screen.getByText(/Create Route/i));
  fireEvent.click(screen.getByRole("button", { name: "acad" }));
  fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  expect(screen.getByLabelText("Waypoint")).toHaveAttribute("disabled");
  expect(screen.getByLabelText("Waypoint")).toHaveValue("Acadia National Park");
  const listItems = screen.getAllByRole("listitem");
  expect(listItems).toHaveLength(1);
  expect(
    screen.getByRole("button", { name: "Remove from Route" })
  ).toBeInTheDocument();

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Remove from Route" }));
  });
  expect(screen.queryAllByLabelText("Waypoint")).toHaveLength(0);
});

test("adding name and notes to the route", async () => {
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const routeName = screen.getByLabelText(/Name/i);
  act(() => {
    fireEvent.click(routeName);
    fireEvent.change(routeName, { target: { value: "Route Name" } });
  });

  const routeNotes = screen.getByLabelText(/Route Notes/i);
  act(() => {
    fireEvent.click(routeNotes);
    fireEvent.change(routeNotes, { target: { value: "Route Notes" } });
  });

  const listItems = screen.getAllByRole("listitem");
  expect(listItems).toHaveLength(2);
  expect(
    screen.getByRole("button", { name: "Save Route" })
  ).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Save Route" })).toHaveAttribute(
    "disabled"
  );
  expect(screen.getByLabelText(/Name/i)).toHaveValue("Route Name");
  expect(screen.getByLabelText(/Route Notes/i)).toHaveValue("Route Notes");
});

test("cancelling the route", async () => {
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const routeName = screen.getByLabelText(/Name/i);
  act(() => {
    fireEvent.click(routeName);
    fireEvent.change(routeName, { target: { value: "Route Name" } });
  });

  const routeNotes = screen.getByLabelText(/Route Notes/i);
  act(() => {
    fireEvent.click(routeNotes);
    fireEvent.change(routeNotes, { target: { value: "Route Notes" } });
  });

  const listItems = screen.getAllByRole("listitem");

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Cancel" }));
  });
  waitFor(() => {
    expect(screen.getByLabelText(/Name/i)).toHaveValue("");
    expect(screen.getByLabelText(/Route Notes/i)).toHaveValue("");
    expect(screen.queryAllByLabelText("Waypoint")).toHaveLength(0);
  });
});

test("saving a route", async () => {
  const user = {
    username: "testUser",
    email: "testUser@gmail.com",
    firstName: "Test",
    lastName: "User",
  };
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_USER", user: user });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const routeName = screen.getByLabelText(/Name/i);
  act(() => {
    fireEvent.click(routeName);
    fireEvent.change(routeName, { target: { value: "My Test Route" } });
  });

  const routeNotes = screen.getByLabelText(/Route Notes/i);
  act(() => {
    fireEvent.click(routeNotes);
    fireEvent.change(routeNotes, { target: { value: "My Test Notes" } });
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Save Route" }));
  });
  await act(async () => await new Promise(process.nextTick));

  expect(screen.getByRole("alert")).toBeInTheDocument();
  expect(screen.getByRole("alert")).toHaveTextContent("Route saved");
  expect(screen.getByLabelText(/Name/i)).toHaveValue("");
  expect(screen.getByLabelText(/Route Notes/i)).toHaveValue("");
  expect(screen.queryAllByLabelText("Waypoint")).toHaveLength(0);
});

test("saved route appears in the My Routes list", async () => {
  const user = {
    username: "testUser",
    email: "testUser@gmail.com",
    firstName: "Test",
    lastName: "User",
  };
  const parks = [
    {
      code: "acad",
      name: "Acadia National Park",
      longitude: "-68.247501",
      latitude: "44.409286",
      parkType: "National Park",
      state: "ME",
    },
    {
      code: "adam",
      name: "Adams National Historical Park",
      longitude: "-71.01160356",
      latitude: "42.2553961",
      parkType: "National Historical Park",
      state: "MA",
    },
  ];

  const store = configureStore({ reducer: root });
  store.dispatch({ type: "FETCH_USER", user: user });
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  await act(async () => await new Promise(process.nextTick));
  store.dispatch({ type: "FETCH_PARKS", parks: parks });
  await act(async () => await new Promise(process.nextTick));

  act(() => {
    fireEvent.click(screen.getByText(/Create Route/i));
    fireEvent.click(screen.getByAltText("acad"));
  });
  act(() => {
    fireEvent.click(screen.getByText("Add to Route"));
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "adam" }));
  });
  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Add to Route" }));
  });

  const routeName = screen.getByLabelText(/Name/i);
  act(() => {
    fireEvent.click(routeName);
    fireEvent.change(routeName, { target: { value: "My Test Route" } });
  });

  const routeNotes = screen.getByLabelText(/Route Notes/i);
  act(() => {
    fireEvent.click(routeNotes);
    fireEvent.change(routeNotes, { target: { value: "My Test Notes" } });
  });

  act(() => {
    fireEvent.click(screen.getByRole("button", { name: "Save Route" }));
  });
  await act(async () => await new Promise(process.nextTick));

  expect(screen.getByRole("alert")).toBeInTheDocument();
  expect(screen.getByRole("alert")).toHaveTextContent("Route saved");
  expect(screen.getByLabelText(/Name/i)).toHaveValue("");
  expect(screen.getByLabelText(/Route Notes/i)).toHaveValue("");
  expect(screen.queryAllByLabelText("Waypoint")).toHaveLength(0);

  act(() => {
    fireEvent.click(screen.getByText(/My Routes/i));
  });
  await act(async () => await new Promise(process.nextTick));

  expect(screen.getByText("My Test Route")).toBeInTheDocument();
});
