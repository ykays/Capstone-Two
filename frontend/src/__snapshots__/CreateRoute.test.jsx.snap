// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateRoute smoke tests > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-1rfid0x-MuiStack-root"
  >
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Create Route
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
